datapath: Add support to delegate routing to other plugins, such as AWS-CNI

Alex-Waring opened the pull request:
AWS uses VLAN interfaces for SG routing, and intentionally avoids upserting a route
so that traffic flows through the VPC and SG rules can be evaluated. Cilium will
upsert the route when an endpoint is created, and this causes traffic to be dropped
due to assymetric routing.

This exposes a wider issue where, when cilium is not in charge of routing traffic,
there is no ability to avoid creating the route (`InstallEndointRoute`) while still
configuring the BPF programs (`RequireEgressProg`). This PR implements a config option
to do just that, meaning we allow the host CNI to configure routing but still enable
network policy for those endpoints (Pods).

This flag is automatically set to True when running with aws-cni, as this is required
due to the issue mentioned above.

I would like this backported to 1.13, however the way to access to cni chaining mode seems to have changed, meaning it may need to be manual.

Fixes: https://github.com/cilium/cilium/issues/27152

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

<!-- Description of change -->

Fixes: https://github.com/cilium/cilium/issues/27152

```release-note
Fix routing delegation to AWS-VPC-CNI when using the security groups feature. 
```

aditighag reviewed (commented):
The latest changes segregating flags look much cleaner. Thanks!
How do you feel about documenting the new flag - https://docs.cilium.io/en/stable/installation/cni-chaining-aws-cni/?

joestringer commented on daemon/cmd/cni/cell.go:
```diff
@@ -69,6 +75,7 @@ func enableConfigManager(lc hive.Lifecycle, log logrus.FieldLogger, cfg Config,
 func newConfigManager(log logrus.FieldLogger, cfg Config, debug bool) *cniConfigManager {
 	if cfg.CNIChainingMode == "aws-cni" && cfg.CNIChainingTarget == "" {
 		cfg.CNIChainingTarget = "aws-cni"
+		cfg.CNIExternalRouting = true
```
@aditighag the nice thing with this particular approach is that for aws-cni we don't need to document anything, it should just autodetect and work due to this line. The existing instructions already configure this option.

Arguably we could document the option further for the general case (non aws-cni), but I think that for now the cmdref should be good enough. We can keep an eye out for any future other external routing agents and point this out if it becomes relevant.

aditighag replied:
Sgtm! 👍

mhofstetter commented on daemon/cmd/cni/cell.go:
```diff
@@ -58,6 +63,7 @@ func (cfg Config) Flags(flags *pflag.FlagSet) {
 	flags.String(option.CNILogFile, defaultConfig.CNILogFile, "Path where the CNI plugin should write logs")
 	flags.String(option.CNIChainingTarget, defaultConfig.CNIChainingTarget, "CNI network name into which to insert the Cilium chained configuration. Use '*' to select any network.")
 	flags.Bool(option.CNIExclusive, defaultConfig.CNIExclusive, "Whether to remove other CNI configurations")
+	flags.Bool(option.CNIExternalRouting, defaultConfig.CNIExternalRouting, "Whether the other CNI handles routing on the node")
```
nit: I would explicitly name _the other_ CNI as the `chained CNI plugin`. It's done the same in the comment of `ExternalRoutingEnabled() bool`

> Whether the other CNI handles routing on the node

->

> Whether the chained CNI plugin handles routing on the node

hint: don't forget to regenerate the CMD ref documentation (`make -C Documentation update-cmdref`)

mhofstetter commented on pkg/option/config.go:
```diff
@@ -1288,6 +1288,9 @@ const (
 	// CNIExclusive tells the agent to remove other CNI configuration files
 	CNIExclusive = "cni-exclusive"
 
+	// CNIExternalRouting delegates endpoint routing to the other CNI.
```
nit: same here. `other CNI` -> `chained CNI plugin`

mhofstetter reviewed (approved):
Thanks! :heavy_check_mark:

I only left a small suggestion regarding the naming of the new flag. Would be nice if you could fix that!

Alex-Waring committed:
datapath: Avoid Upserting AWS SG vlans

When cilium is not in charge of routing traffic,
there is no ability to avoid creating the route (`InstallEndointRoute`) while still
configuring the BPF programs (`RequireEgressProg`). This PR implements a config option
to do just that, meaning we allow the host CNI to configure routing but still enable
network policy for those endpoints. This is defaulted to enabled for aws-cni chaining.

Fixes: #27152
Signed-off-by: Alex Waring <ajmwaring@gmail.com>
Co-authored-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Joe Stringer <joe@cilium.io>

sayboras reviewed (approved):
Thanks 🍒
